airflow/executors/celery_executor.py

```diff
@@ -469,6 +505,15 @@ def try_adopt_task_instances(self, tis: List[TaskInstance]) -> List[TaskInstance
 
         return not_adopted_tis
 
+    def _set_stalled_task_timeout(self, key: TaskInstanceKey, timeout: datetime.timedelta) -> None:
+        if timeout:
+            self.stalled_task_timeouts[key] = utcnow() + timeout
+            self.stalled_task_timeouts = dict(
+                sorted(self.stalled_task_timeouts.items(), key=lambda item: item[1])
+            )
```

Comment:
Hmmm, the performance implications of re-sorting this _every time_ we add one task is not great (n^2 complexity)

We should look at doing this better -- either find a data structure that stores this sorted on insert (a "heap") or only do this sorting once at the end, not on every write.

🔋 https://docs.python.org/3/library/heapq.html

Reply:
After shifting the adopted tasks and stalled tasks to each be tracked in their own dicts, sorting is no longer necessary at all as both dicts inherently stay sorted naturally